### PR TITLE
Generate keys with metadata

### LINKED ISSUE #51: Move KeyGenerator to be in an enveloping class
See https://github.com/SSH-key-rotation-AWS/team-henrique/pull/46#discussion_r1247888426

### LINKED ISSUE #58: Missing key metadata
Currently keys are generated alone - should be generated with a metadata JSON file instead.

### PR DESCRIPTION
**Pull request by ygins**
- Added KeyMetadata class with created_by and time_created fields
 - Added methods to the publishers to publish metadata
 - Started migrating some code to PathLib, still a WIP
 - Added tests for S3 publisher and metadata

closes #58 
closes #51

### CODE REVIEW

**Comment by ygins** [comment]
closes #58 
closes #51

**Review by yydiamen — APPROVED** [review]

File: switcheroo/server/retrieve_public_keys.py [hunk 1]
```diff
@@ -7,8 +7,8 @@
 
 def get_public_keys_local(connecting_user: str, ssh_home_dir: str) -> str:
     host_name = socket.getfqdn()
-    key_dir = f"{ssh_home_dir}/{host_name}"
-    key_path = f"{key_dir}/{connecting_user}-cert.pub"
+    key_dir = f"{ssh_home_dir}/{host_name}/{connecting_user}"
+    key_path = f"{key_dir}/key-cert.pub"
     if not os.path.isdir(key_dir):
```
Comment (thehcma) [inline switcheroo/server/retrieve_public_keys.py:12]: I feel like I reviewed this already and we didn't act on the review - so here it goes again - don't test and act - rather act and handle the exception

also use `pathlib`

File: switcheroo/server/retrieve_public_keys.py [hunk 2]
```diff
@@ -21,7 +21,7 @@ def get_public_keys_s3(connecting_user: str, bucket_name: str) -> str:
     host_name = socket.getfqdn()
     s3_client = boto3.client("s3")
```
Comment (thehcma) [inline switcheroo/server/retrieve_public_keys.py:22]: this should be done via the datastore (using AWS should be done only when the AWS S3 data store is in use)

**Review by thehcma — CHANGES_REQUESTED** [review]

File: switcheroo/custom_keygen.py [hunk 3]
```diff
@@ -1,51 +1,141 @@
...
+
+        if not os.path.isdir(private_key_dir):
+            os.makedirs(private_key_dir)
+        if not os.path.isdir(public_key_dir):
+            os.makedirs(public_key_dir)
+
+        private_key_path = f"{private_key_dir}/{private_key_name}"
+        public_key_path = f"{public_key_dir}/{public_key_name}"
+
+        with open(private_key_path, "wb") as private_out:
```
[the remark this answers is not in the document]
Comment (ygins) [reply switcheroo/custom_keygen.py:133]: pathlib.Path.mkdir takes a mode - can we do that instead?

[the remark this answers is not in the document]
Comment (thehcma) [reply switcheroo/custom_keygen.py:133]: `mkdir` creates a directory - you need to create the file with the proper permissions as well

**Pull request merged**